### PR TITLE
Use `ThinVec` for PredicateObligation storage

### PR DESCRIPTION
**Pull request by GnomedDev**
~~I noticed while profiling clippy on a project that a large amount of time is being spent allocating `Vec`s for `PredicateObligation`, and the `Vec`s are often quite small. This is an attempt to optimise this by using SmallVec to avoid heap allocations for these common small Vecs.~~ 

This PR turns all the `Vec<PredicateObligation>` into a single type alias while avoiding referring to `Vec` around it, then swaps the type over to `ThinVec<PredicateObligation>` and fixes the fallout. This also contains an implementation of `ThinVec::extract_if`, copied from `Vec::extract_if` and currently being upstreamed to https://github.com/Gankra/thin-vec/pull/66.

This leads to a small (0.2-0.7%) performance gain in the latest perf run.

### CODE REVIEW

**Comment by workingjubilee** [comment]
it seems branching on every access is much more costly than doing an unconditional dereference.

**Comment by workingjubilee** [comment]
that, or copying the thinner data structures around saves much more than saving a load to check lens.

**Comment by GnomedDev** [comment]
Keep in mind, ThinVec also probably has some easy wins in improving it's implementation as seen by my capacity overflow PR to it. This opens the door for further performance wins. I'll get this PR cleaned up and reviewable and hopefully get this merged in.

File: compiler/rustc_trait_selection/src/solve/fulfill.rs [hunk 1]
```diff
@@ -81,7 +82,7 @@ impl<'tcx> ObligationStorage<'tcx> {
             // we get all obligations involved in the overflow. We pretty much check: if
             // we were to do another step of `select_where_possible`, which goals would
             // change.
-            self.overflowed.extend(self.pending.extract_if(|o| {
+            self.overflowed.extend(ExtractIf::new(&mut self.pending, |o| {
```
[the remark this answers is not in the document]
Comment (GnomedDev) [reply compiler/rustc_trait_selection/src/solve/fulfill.rs:86]: Done!

**Comment by GnomedDev** [comment]
Ready to review!

@rustbot review
r? compiler

**Comment by fee1-dead** [comment]
r? compiler

**Review by compiler-errors — APPROVED** [review]

**Comment by Kobzol** [comment]
Nice results! :) Good find.

**Comment by pnkfelix** [comment]
Visiting for perf triage

* improvements outweigh regressions. (and reported performance matches what was anticipated via perf runs.)
* marking as triaged

@rustbot label: +perf-regression-triaged

**Pull request merged**